app/services/projects/copy/work_packages_dependent_service.rb

```diff
@@ -0,0 +1,157 @@
+        service_call.result
+      else
+        add_error!(source_work_package, service_call.errors)
+        error = service_call.message
+        Rails.logger.warn do
+          "Project#copy_work_packages: work package ##{source_work_package.id} could not be copied: #{error}"
+        end
+
+        nil
+      end
+    end
+
+    def copy_relations(wp, new_wp_id, work_packages_map)
+      wp.relations_to.non_hierarchy.direct.each do |source_relation|
+        new_relation = Relation.new
+        new_relation.attributes = source_relation.attributes.dup.except('id', 'from_id', 'to_id', 'relation_type')
+        new_relation.to_id = work_packages_map[source_relation.to_id]
+        if new_relation.to_id.nil? && Setting.cross_project_work_package_relations?
+          new_relation.to_id = source_relation.to_id
+        end
+        new_relation.from_id = new_wp_id
+        new_relation.save
+      end
+
+      wp.relations_from.non_hierarchy.direct.each do |source_relation|
+        new_relation = Relation.new
+        new_relation.attributes = source_relation.attributes.dup.except('id', 'from_id', 'to_id', 'relation_type')
+        new_relation.from_id = work_packages_map[source_relation.from_id]
+        if new_relation.from_id.nil? && Setting.cross_project_work_package_relations?
+          new_relation.from_id = source_relation.from_id
+        end
+        new_relation.to_id = new_wp_id
+        new_relation.save
+      end
+
+    end
+
+    def copy_work_package_attribute_overrides(source_work_package, parent_id, user_cf_ids)
+      custom_value_attributes = source_work_package.custom_value_attributes.map do |id, value|
+        if user_cf_ids.include?(id) && !target.users.detect { |u| u.id.to_s == value }
+          [id, nil]
+        else
+          [id, value]
+        end
+      end.to_h
+
+      {
+        project: target,
+        parent_id: parent_id,
+        version: work_package_version(source_work_package),
+        assigned_to: work_package_assigned_to(source_work_package),
+        responsible: work_package_responsible(source_work_package),
+        custom_field_values: custom_value_attributes,
+        # We fetch the value from the global registry to persist it in the job which
+        # will trigger a delayed job for potentially sending the journal notifications.
+        send_notifications: ActionMailer::Base.perform_deliveries
+      }
+    end
+
+    def work_package_version(source_work_package)
+      source_work_package.version && target.versions.detect { |v| v.name == source_work_package.version.name }
+    end
+
+    def work_package_assigned_to(source_work_package)
+      source_work_package.assigned_to && target.possible_assignees.detect { |u| u.id == source_work_package.assigned_to_id }
+    end
+
+    def work_package_responsible(source_work_package)
+      source_work_package.responsible && target.possible_responsibles.detect { |u| u.id == source_work_package.responsible_id }
+    end
```

Comment:
Those lookup will run repeatedly, for every work package to copy. Transforming it into a hash that is kept between lookups would be way faster.

Reply:
For versions, we can even use a lookup map as they got created before. Will do that